File: src/ValidPalindrome125.java
public class ValidPalindrome125 {
    public boolean isPalindrome(String s) {
        int p1 = 0, p2 = s.length() - 1;

        while (p1 < p2) {
            while (p1 < p2 && !Character.isLetterOrDigit(s.charAt(p1))) p1++;
            while (p2 > p1 && !Character.isLetterOrDigit(s.charAt(p2))) p2--;

            if (Character.toLowerCase(s.charAt(p1)) != Character.toLowerCase(s.charAt(p2)))
                return false;

            p1++;
            p2--;
        }

        return true;
    }

    public static void main(String[] args) {
        ValidPalindrome125Funs s = new ValidPalindrome125Funs();

        System.out.println(s.isPalindrome("asdsa"));
        System.out.println(s.isPalindrome("asdff"));
        System.out.println(s.isPalindrome("A\tsd  Sa"));
    }
}

File: src/ValidPalindrome125.kt
/**
 * https://leetcode.com/problems/valid-palindrome
 */
class ValidPalindrome125 {
    fun isPalindrome(s: String): Boolean {
        val res = s.toLowerCase().filter { it.isLetterOrDigit() }

        return res == res.reversed()
    }
}

class ValidPalindrome125TwoPointer {
    fun isPalindrome(s: String): Boolean {
        var p1 = 0
        var p2 = s.length - 1

        while (p1 < p2) {
            while (p1 < p2 && s[p1].isLetterOrDigit().not()) p1++
            while (p2 > p1 && s[p2].isLetterOrDigit().not()) p2--

            if (s[p1].toLowerCase() != s[p2].toLowerCase())
                return false

            p1++
            p2--
        }

        return true
    }
}

fun main() {
    println(ValidPalindrome125TwoPointer().isPalindrome("A man, a plan, a canal: Panama"));
    println(ValidPalindrome125TwoPointer().isPalindrome("race a car"));
}